File: python/330_按要求补齐数组_hard.py
class Solution:
    def minPatches(self, nums: List[int], n: int) -> int:
        length = len(nums)
        index, ans = 0, 0
        x = 1
        while x <= n:
            if index < length and nums[index] <= x:
                x += nums[index]
                index += 1
            else:
                x <<= 1
                ans += 1
        return ans

File: cpp/330_按要求补齐数组_hard.cpp
class Solution {
public:
    int minPatches(vector<int>& nums, int n) {
        int length = nums.size();
        long long x = 1;
        int index = 0;
        int ans = 0;
        while(x<=n)
            if(index < length && nums[index] < x+1){
                x += nums[index];
                index++;
            }
            else{
                x <<= 1;
                ans ++;
            }
        return ans;
    }
};